src/components/MagicCodeInput.js

```diff
@@ -0,0 +1,305 @@
+                            ...numbers.slice(0, focusedIndex - 1),
+                            CONST.MAGIC_CODE_EMPTY_CHAR,
+                            ...numbers.slice(focusedIndex, CONST.MAGIC_CODE_LENGTH),
+                        ],
+                    focusedIndex: Math.max(0, focusedIndex - 1),
+                    editIndex: Math.max(0, focusedIndex - 1),
+                };
+            });
+        } else if (keyValue === 'ArrowLeft') {
+            this.setState(prevState => ({
+                input: '',
+                focusedIndex: prevState.focusedIndex - 1,
+                editIndex: prevState.focusedIndex - 1,
+            }));
+        } else if (keyValue === 'ArrowRight') {
+            this.setState(prevState => ({
+                input: '',
+                focusedIndex: prevState.focusedIndex + 1,
+                editIndex: prevState.focusedIndex + 1,
+            }));
+        } else if (keyValue === 'Enter') {
+            this.setState({input: ''});
+            this.props.onFulfill(this.composeToString(this.state.numbers));
+        }
+    }
+
+    /**
+     * Converts a given string into an array of numbers that must have the same
+     * number of elements as the number of inputs.
+     *
+     * @param {String} value
+     * @returns {Array}
+     */
+    decomposeString(value) {
+        let arr = _.map(value.split('').slice(0, CONST.MAGIC_CODE_LENGTH), v => (ValidationUtils.isNumeric(v) ? v : CONST.MAGIC_CODE_EMPTY_CHAR));
+        if (arr.length < CONST.MAGIC_CODE_LENGTH) {
+            arr = arr.concat(Array(CONST.MAGIC_CODE_LENGTH - arr.length).fill(CONST.MAGIC_CODE_EMPTY_CHAR));
+        }
+        return arr;
+    }
+
+    /**
+     * Converts an array of strings into a single string. If there are undefined or
+     * empty values, it will replace them with a space.
+     *
+     * @param {Array} value
+     * @returns {String}
+     */
+    composeToString(value) {
+        return _.map(value, v => ((v === undefined || v === '') ? CONST.MAGIC_CODE_EMPTY_CHAR : v)).join('');
+    }
+
+    render() {
+        return (
+            <>
+                <View style={[styles.flexRow, styles.justifyContentBetween]}>
+                    {_.map(this.inputPlaceholderSlots, index => (
+                        <View key={index} style={[styles.w15]}>
+                            <View
+                                style={[
+                                    styles.textInputContainer,
+                                    this.state.focusedIndex === index ? styles.borderColorFocus : {},
+                                ]}
+                            >
+                                <Text style={[styles.magicCodeInput, styles.textAlignCenter]}>
+                                    {this.state.numbers[index] || ''}
+                                </Text>
+                            </View>
+                            <View style={[StyleSheet.absoluteFillObject, styles.w100, styles.opacity0]}>
```

Comment:
> I don't like this style definition

This is necessary to hide the input above the text, so that we can keep that behaviour of copy paste in each input as well as focus on tap.
What exactly don't you like abou the style definition? Do you want me to move it to a new class?

> Actually why can't we use TextInput instead of manual Text?

This is something I had previously, but it was too problematic. You would need to change the focus to the input above and force it to never focus, pass a `focused` prop (which was another prop needed to add to the BaseTextInput only to show the focused state), be non editable. With Text it becomes simpler, it's only a wrapper for a number with some style.